File: day06/solution.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

type Coord struct {
	row int
	col int
}
type CoordAndDir struct {
	row int
	col int
	dir int
}

func main(){
	grid:=readData("input.txt")
	// fmt.Println(grid)
	startCoord:=findChar(grid,"^")
	part1(grid,startCoord)
	part2(grid,startCoord)
}

func part2(grid[][]string,startCoord Coord){
	rowCount:=len(grid)
	colCount:=len(grid[0])
	coord:=startCoord
	direction:=0
	guardPath:=map[Coord]bool{}
	for true{
		nextCoord:=move(coord,direction)
		if inBounds(nextCoord,rowCount,colCount){
			// on-map
			switch grid[nextCoord.row][nextCoord.col]{
				case "#":
					// box: turn right,
					direction=(direction+1)%4
				case "^":
					// guard start. advance but don't add to guardPath
					coord=nextCoord
				default:
					// path clear, move
					guardPath[nextCoord]=true;
					coord=nextCoord
			}
		} else {
			// off map
			break
		}
	}

	loopCount:=0
	for boxCoord:=range guardPath{
		// deep copy so that updates dont corrupt the grid
		var testGrid[][]string
		for _,row:=range grid{
			testGrid=append(testGrid,append([]string(nil),row...))
		}
		testGrid[boxCoord.row][boxCoord.col]="#"
		testPath:=map[CoordAndDir]bool{}
		coord:=startCoord
		direction:=0
		// fmt.Println("drop box at",boxCoord,"loops:",loopCount)

		for true{
			nextCoord:=move(coord,direction)
			if inBounds(nextCoord,rowCount,colCount){
				// on-map
				if testGrid[nextCoord.row][nextCoord.col]=="#" {
					// box: turn right,
					direction=(direction+1)%4
				} else {
					// path clear, move
					nextCoordAndDir:=CoordAndDir{nextCoord.row,nextCoord.col,direction}
					foundPath,_:=testPath[nextCoordAndDir]
					if foundPath{
						loopCount++
						fmt.Println("loop",loopCount,"at box",boxCoord)
						break
					}
					testPath[nextCoordAndDir]=true;
					coord=nextCoord
				}
			} else {
				// off map
				break
			}
		}
	}
	// BUG: off by one - should be 1719, get 1720
	fmt.Println("part 2:",loopCount,"len guardPath",len(guardPath))
}

func part1(grid[][]string,startCoord Coord){
	rowCount:=len(grid)
	colCount:=len(grid[0])
	coord:=startCoord
	direction:=0
	distinctCoords:=map[Coord]bool{}
	for true{
		nextCoord:=move(coord,direction)
		if inBounds(nextCoord,rowCount,colCount){
			// on-map
			if grid[nextCoord.row][nextCoord.col]=="#" {
				// box: turn right,
				direction=(direction+1)%4
			} else {
				// path clear, move
				distinctCoords[nextCoord]=true;
				coord=nextCoord
			}
		} else {
			// off map
			break
		}
	}
	fmt.Printf("part 1: %d\n",len(distinctCoords))
}

func inBounds(coord Coord,rowCount int,colCount int)(bool){
	return 0<=coord.row && coord.row<rowCount && 0<=coord.col && coord.col<colCount
}

func move(coord Coord,dir int)(Coord){
	newCoord:=Coord{row:coord.row,col:coord.col}
	switch dir{
		case 0: newCoord.row--
		case 1: newCoord.col++
		case 2: newCoord.row++
		case 3: newCoord.col--
	}
	return newCoord
}

func findChar(grid[][]string,find string)(Coord){
	for rowIx,row:=range grid{
		for colIx,char:=range row{
			if char==find{
				return Coord{row:rowIx,col:colIx}
			}
		}
	}
	return Coord{row:-1,col:-1}
}

func readData(filename string)([][]string){
	f,_:=os.Open(filename)
	defer f.Close()
	var data[][]string
	scanner:=bufio.NewScanner(f)
	for scanner.Scan(){
		line:=scanner.Text()
		chars:=strings.Split(line,"")
		data=append(data, chars)
	}
	return data
}

File: day06/solution.cpp
#include <cassert>
#include <cstddef>
#include <cstdlib>
#include <fstream>
#include <functional>
#include <iostream>
#include <set>
#include <stdexcept>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

struct Args {
  std::string filename = "example.txt";
  int debug = 0;
  bool part1 = true;
  bool part2 = true;
};

struct TextGrid {
  std::string data = "";
  int rows = 0;
  int cols = 0;
  int boxRow = -1;
  int boxCol = -1;
  char at(int row, int col) {
    if (oob(row, col))
      return ' ';
    if (row == boxRow && col == boxCol)
      return '#';
    return data.at(row * cols + col);
  }
  std::pair<int, int> find(char c) {
    // for finding start/end pos
    const int ix = data.find(c);
    return {ix / cols, ix % cols};
  }
  bool oob(int row, int col) {
    return row < 0 || row >= rows || col < 0 || col >= cols;
  };
  // lightweight alternative to creating a new instance and copying/replacing
  // string data
  void box(int row = -1, int col = -1) { boxRow = row, boxCol = col; }
};

struct RowColDir {
  int r, c, d;
  // so the struct can be added to a set
  bool operator<(const RowColDir &rhs) const {
    return r != rhs.r ? r < rhs.r : c != rhs.c ? c < rhs.c : d < rhs.d;
  }
  bool operator=(const RowColDir &rhs) const {
    return r == rhs.r && c == rhs.c && d == rhs.d;
  }
};

Args parseArgs(int argc, char *argv[]) {
  Args args;
  std::unordered_map<std::string, std::function<void()>> argMap = {
      {"-i", [&]() { args.filename = "input.txt"; }},
      {"-d", [&]() { args.debug = 1; }},
      {"-d1", [&]() { args.debug = 1; }},
      {"-d2", [&]() { args.debug = 2; }},
      {"-d3", [&]() { args.debug = 3; }},
      {"-p1",
       [&]() {
         args.part1 = true;
         args.part2 = false;
       }},
      {"-p2",
       [&]() {
         args.part1 = false;
         args.part2 = true;
       }},
  };
  for (int i = 1; i < argc; ++i) {
    std::string arg = argv[i];
    if (arg.rfind("-e", 0) == 0 && arg.size() > 2) {
      args.filename = "example" + arg.substr(2) + ".txt";
    } else if (argMap.find(arg) != argMap.end()) {
      argMap[arg](); // Execute the associated action
    } else {
      throw std::runtime_error("unknown arg: " + arg);
    }
  }
  std::cout << "filename: " << args.filename << "; debug: " << args.debug
            << "; part 1: " << std::boolalpha << args.part1
            << "; part 2: " << std::boolalpha << args.part2 << "\n";
  return args;
}

TextGrid readData(std::string filename, int debug) {
  TextGrid grid;
  std::ifstream file(filename);
  if (file.is_open()) {
    std::string line;
    int rows = 0;
    while (std::getline(file, line)) {
      grid.data += line;
      ++rows;
      if (grid.cols == 0) {
        grid.cols = line.size();
      }
    }
    grid.rows = rows;
    file.close();
  } else {
    throw std::runtime_error("cannot open: " + filename);
  }
  if (debug > 1) {
    std::cout << "grid rows: " << grid.rows << "; cols: " << grid.cols
              << "; data size:" << grid.data.size()
              << "; at(0,0): " << grid.at(0, 0) << "\n";
  }
  return grid;
}

void part1(TextGrid &grid, int debug) {
  auto [r, c] = grid.find('^');
  std::array<std::pair<int, int>, 4> dirs = {{
      {-1, 0},
      {0, 1},
      {1, 0},
      {0, -1},
  }};
  int d = 0;
  if (debug > 0)
    std::cout << "start r: " << r << " c: " << c << " d: " << d << "\n";
  std::set<std::pair<int, int>> locations;
  while (!grid.oob(r, c)) {
    int nr = r + dirs[d].first, nc = c + dirs[d].second;
    if (debug > 0) {
      std::cout << "r: " << r << " c: " << c << " d: " << d << " nr: " << nr
                << " nc: " << nc << " locs: " << locations.size() << "\n";
    }
    // turn right at box
    if (grid.at(nr, nc) == '#') {
      d = (d + 1) % 4;
      continue;
    }
    // add to locations set if not oob
    if (!grid.oob(nr, nc)) {
      locations.insert({nr, nc});
    }
    // walk (oob will break the loop)
    r = nr, c = nc;
  };
  std::cout << "part 1: " << locations.size() << "\n";
}

void part2(TextGrid &grid, int debug) {
  auto start = grid.find('^');
  std::array<std::pair<int, int>, 4> dirs = {{
      {-1, 0},
      {0, 1},
      {1, 0},
      {0, -1},
  }};
  int d = 0, r = start.first, c = start.second;
  std::set<std::pair<int, int>> boxLocations;
  // initial walk to build boxLocations
  while (!grid.oob(r, c)) {
    int nr = r + dirs[d].first, nc = c + dirs[d].second;
    if (grid.at(nr, nc) == '#') {
      d = (d + 1) % 4;
      continue;
    }
    if (!grid.oob(nr, nc))
      boxLocations.insert({nr, nc});
    r = nr, c = nc;
  };
  int loopCount = 0;
  // boxLocations.clear();
  // // boxLocations.insert({55, 85});
  boxLocations.insert({55, 86});
  for (auto box : boxLocations) {
    // DO NOT THE START
    if (box.first == start.first && box.second == start.second)
      continue;
    debug = (box.first == 55 && box.second == 86) ? 3 : 0;
    // add a pseudo box to grid
    grid.box(box.first, box.second);
    // store a history of r,c,d so we can determine if we're in a loop
    std::set<RowColDir> boxPath;
    // reset to start
    int d = 0, r = start.first, c = start.second;
    boxPath.insert({r, c, d});
    if (debug > 0)
      std::cout << "box=(" << box.first << "," << box.second
                << ") boxGrid.at=" << grid.at(box.first, box.second) << "\n";
    while (!grid.oob(r, c)) {
      int nr = r + dirs[d].first, nc = c + dirs[d].second;
      // turn or advance
      if (grid.at(nr, nc) == '#') {
        d = (d + 1) % 4;
        if (debug > 1)
          std::cout << "  hit box at (" << nr << "," << nc << ") new d=" << d
                    << " r=" << r << " c=" << c << "\n";
      } else {
        r = nr, c = nc;
      }
      // loop check
      const auto rcd = RowColDir{r, c, d};
      if (boxPath.find(rcd) != boxPath.end()) {
        if (debug > 0)
          std::cout << "  loop: box=(" << box.first << "," << box.second
                    << ") nr=" << nr << " nc=" << nc << " d=" << d << "\n";
        ++loopCount;
        break;
      }
      // add new rcd to path
      boxPath.insert(rcd);
    }
    if (debug > 0)
      std::cout << "  end d=" << d << " r=" << r << " c=" << c << "\n";
  }
  /*
    fixed bug:
      start position was correctly ommitted from boxLocations on create but was
    added during initial walk (albeit in a different direction)
      it is invalid to put a box at the start position, but it does result in a
    loop if you don't check for it before the walk
  */
  std::cout << "part 2: " << loopCount << "\n";
}

int main(int argc, char *argv[]) {
  auto args = parseArgs(argc, argv);
  auto grid = readData(args.filename, args.debug);
  if (args.part1)
    part1(grid, args.debug);
  if (args.part2)
    part2(grid, args.debug);
  return 0;
}